Tests/SourceKitLSPTests/SwiftInterfaceTests.swift

```diff
@@ -0,0 +1,154 @@
+  
+  func testSystemModuleInterface() throws {
+    let url = URL(fileURLWithPath: "/\(UUID())/a.swift")
+    let uri = DocumentURI(url)
+
+    sk.send(DidOpenTextDocumentNotification(textDocument: TextDocumentItem(
+      uri: uri,
+      language: .swift,
+      version: 1,
+      text: """
+      import Foundation
+      """)))
+    
+    let _resp = try sk.sendSync(DefinitionRequest(
+      textDocument: TextDocumentIdentifier(url),
+      position: Position(line: 0, utf16index: 10)))
+    let resp = try XCTUnwrap(_resp)
+    guard case .locations(let locations) = resp else {
+      XCTFail("Unexpected response: \(resp)")
+      return
+    }
+    XCTAssertEqual(locations.count, 1)
+    let location = try XCTUnwrap(locations.first)
+    XCTAssertTrue(location.uri.pseudoPath.hasSuffix("/Foundation.swiftinterface"))
+    let fileContents = try XCTUnwrap(location.uri.fileURL.flatMap({ try String(contentsOf: $0, encoding: .utf8) }))
+    XCTAssertTrue(fileContents.hasPrefix("import "))
+  }
+  
+  func testOpenInterface() throws {
+    guard let ws = try staticSourceKitSwiftPMWorkspace(name: "SwiftPMPackage") else { return }
+    try ws.buildAndIndex()
+    let call = ws.testLoc("Lib.foo:call")
+    try ws.openDocument(call.url, language: .swift)
+    let workspace = try XCTUnwrap(ws.testServer.server?.queue.sync {
+      try XCTUnwrap(ws.testServer.server?.workspaceForDocument(uri: call.docUri))
+    })
+    let swiftLangServer = try XCTUnwrap(ws.testServer.server?._languageService(for: call.docUri,
+                                                                               .swift,
+                                                                               in: workspace))
+    let expectation = expectation(description: "open interface request")
+    let openInterface = OpenInterfaceRequest(textDocument: call.docIdentifier, name: "lib")
+    let request = Request(openInterface, id: .number(1), clientID: ObjectIdentifier(swiftLangServer),
+                          cancellation: CancellationToken(), reply: { (result: Result<OpenInterfaceRequest.Response, ResponseError>) in
+      do {
+        let interfaceDetails = try result.get()
+        XCTAssertTrue(interfaceDetails.uri.pseudoPath.hasSuffix("/lib.swiftinterface"))
+        let fileContents = try XCTUnwrap(interfaceDetails.uri.fileURL.flatMap({ try String(contentsOf: $0, encoding: .utf8) }))
+        XCTAssertTrue(fileContents.contains("""
+          public struct Lib {
+          
+              public func foo()
+          
+              public init()
+          }
+          """))
+      } catch {
+        XCTFail(error.localizedDescription)
+      }
+      expectation.fulfill()
+    })
+    
+    // Send an arbitrary request through the front door first or SourceKitServer won't be properly initialized. 
+    _ = try ws.sk.sendSync(HoverRequest(
+      textDocument: call.docIdentifier,
+      position: Position(line: 0, utf16index: 8)))
+    swiftLangServer.openInterface(request)
```

Comment:
Is there any reason why this can’t be `ws.sk.sendSync(openInterface)`. If you send the request that way, I think you also don’t need to send the hover request.

Reply:
If I do it that way it hits this `fatalError`
https://github.com/apple/sourcekit-lsp/blob/main/Sources/LSPTestSupport/TestJSONRPCConnection.swift#L145

Reply:
Oh, that’s a very important catch that would have prevented the open interface request from working in the live `sourcekit-lsp` server as well: You need to add `OpenInterfaceRequest.self` to `builtinRequests` in `Messages.swift`.

Reply:
I added `OpenInterfaceRequest.self` to `builtinRequests` in `Messages.swift` but still get the same crash.

Reply:
I just did a little bit more debugging myself and I think these are the changes you need. Essentially, what needs to be done is that SourceKit-LSP has to know about the request type and which method implements the request handling (`registerToolchainTextDocumentRequest`). That also needs to have a fallback response, that's why I changed the return type to be optional.

```diff
diff --git a/Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift b/Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift
index db33f31..6292c1d 100644
--- a/Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift
+++ b/Sources/LanguageServerProtocol/Requests/OpenInterfaceRequest.swift
@@ -14,7 +14,7 @@
 /// **(LSP Extension)**
 public struct OpenInterfaceRequest: TextDocumentRequest, Hashable {
   public static let method: String = "textDocument/openInterface"
-  public typealias Response = InterfaceDetails
+  public typealias Response = InterfaceDetails?
 
   /// The document whose compiler arguments should be used to generate the interface.
   public var textDocument: TextDocumentIdentifier
diff --git a/Sources/SourceKitLSP/SourceKitServer.swift b/Sources/SourceKitLSP/SourceKitServer.swift
index 28f93a7..ef07a3e 100644
--- a/Sources/SourceKitLSP/SourceKitServer.swift
+++ b/Sources/SourceKitLSP/SourceKitServer.swift
@@ -184,6 +184,7 @@ public final class SourceKitServer: LanguageServer {
     registerToolchainTextDocumentRequest(SourceKitServer.completion,
                                          CompletionList(isIncomplete: false, items: []))
     registerToolchainTextDocumentRequest(SourceKitServer.hover, nil)
+    registerToolchainTextDocumentRequest(SourceKitServer.openInterface, nil)
     registerToolchainTextDocumentRequest(SourceKitServer.declaration, .locations([]))
     registerToolchainTextDocumentRequest(SourceKitServer.definition, .locations([]))
     registerToolchainTextDocumentRequest(SourceKitServer.references, [])
@@ -986,6 +987,14 @@ extension SourceKitServer {
     languageService.hover(req)
   }
 
+  func openInterface(
+    _ req: Request<OpenInterfaceRequest>,
+    workspace: Workspace,
+    languageService: ToolchainLanguageServer
+  ) {
+    languageService.openInterface(req)
+  }
+
   /// Find all symbols in the workspace that include a string in their name.
   /// - returns: An array of SymbolOccurrences that match the string.
   func findWorkspaceSymbols(matching: String) -> [SymbolOccurrence] {
@@ -1279,9 +1288,11 @@ extension SourceKitServer {
         let request = Request(openInterface, id: req.id, clientID: ObjectIdentifier(self),
                               cancellation: req.cancellationToken, reply: { (result: Result<OpenInterfaceRequest.Response, ResponseError>) in
           switch result {
-          case .success(let interfaceDetails):
+          case .success(let interfaceDetails?):
             let loc = Location(uri: interfaceDetails.uri, range: Range(Position(line: 0, utf16index: 0)))
             req.reply(.locations([loc]))
+          case .success(nil):
+            req.reply(.failure(.unknown("Could not generate Swift Interface for \(name)")))
           case .failure(let error):
             req.reply(.failure(error))
           }
diff --git a/Tests/SourceKitLSPTests/SwiftInterfaceTests.swift b/Tests/SourceKitLSPTests/SwiftInterfaceTests.swift
index df39e33..44da807 100644
--- a/Tests/SourceKitLSPTests/SwiftInterfaceTests.swift
+++ b/Tests/SourceKitLSPTests/SwiftInterfaceTests.swift
@@ -91,35 +91,19 @@ final class SwiftInterfaceTests: XCTestCase {
       try XCTUnwrap(ws.testServer.server?.workspaceForDocument(uri: importedModule.docUri))
     })
     let swiftLangServer = try XCTUnwrap(ws.testServer.server?._languageService(for: importedModule.docUri, .swift, in: workspace))
-    let expectation = expectation(description: "open interface request")
     let openInterface = OpenInterfaceRequest(textDocument: importedModule.docIdentifier, name: "lib")
-    let request = Request(openInterface, id: .number(1), clientID: ObjectIdentifier(swiftLangServer),
-                          cancellation: CancellationToken(), reply: { (result: Result<OpenInterfaceRequest.Response, ResponseError>) in
-      do {
-        let interfaceDetails = try result.get()
-        XCTAssertTrue(interfaceDetails.uri.pseudoPath.hasSuffix("/lib.swiftinterface"))
-        let fileContents = try XCTUnwrap(interfaceDetails.uri.fileURL.flatMap({ try String(contentsOf: $0, encoding: .utf8) }))
-        XCTAssertTrue(fileContents.contains("""
-          public struct Lib {
-          
-              public func foo()
-          
-              public init()
-          }
-          """))
-      } catch {
-        XCTFail(error.localizedDescription)
-      }
-      expectation.fulfill()
-    })
-    
-    // Send an arbitrary request through the front door first or SourceKitServer won't be properly initialized.
-    _ = try ws.sk.sendSync(HoverRequest(
-      textDocument: importedModule.docIdentifier,
-      position: importedModule.position))
-    swiftLangServer.openInterface(request)
     
-    waitForExpectations(timeout: 15)
+    let interfaceDetails = try XCTUnwrap(ws.sk.sendSync(openInterface))
+    XCTAssertTrue(interfaceDetails.uri.pseudoPath.hasSuffix("/lib.swiftinterface"))
+    let fileContents = try XCTUnwrap(interfaceDetails.uri.fileURL.flatMap({ try String(contentsOf: $0, encoding: .utf8) }))
+    XCTAssertTrue(fileContents.contains("""
+      public struct Lib {
+
+          public func foo()
+
+          public init()
+      }
+      """))
   }
 
   func testSwiftInterfaceAcrossModules() throws {
```

Reply:
Thanks for looking into that! I've made the recommended changes.